File: frontend/src/services/api/ingestion.ts
/**
 * ARCH-38 — batch ingestion API client.
 *
 * `uploadFileResumable` is the only place in the frontend that knows how a
 * resumable upload works. It slices the file at the part size the SERVER
 * chose, and before sending anything it asks the server which parts it already
 * holds. That single call is what makes a page reload survivable: the browser
 * keeps no record of its own, so there is nothing to go stale.
 */

import type { AxiosProgressEvent } from "axios";

import apiClient from "@/services/api/client";
import {
  INGESTION_ENDPOINTS,
  WORK_ITEM_ENDPOINTS,
} from "@/services/api/endpoints";

const JSON_HEADERS = { Accept: "application/json" } as const;

export interface BatchFileInput {
  readonly client_key: string;
  readonly filename: string;
  readonly size_bytes: number;
  readonly sha256?: string;
}

export interface BatchItem {
  readonly id: string;
  readonly client_key: string;
  readonly filename: string;
  readonly size_bytes: number;
  readonly status: string;
  readonly error_code?: string | null;
  readonly error_detail?: string | null;
  readonly work_item_id?: string | null;
  readonly upload_session_id?: string | null;
}

export interface Batch {
  readonly id: string;
  readonly status: string;
  readonly source: string;
  readonly total_items: number;
  readonly completed_items: number;
  readonly failed_items: number;
  readonly created_at: string;
  readonly completed_at?: string | null;
  readonly items: readonly BatchItem[];
}

export interface UploadSession {
  readonly id: string;
  readonly filename: string;
  readonly part_size: number;
  readonly parts_received: readonly number[];
  readonly status: string;
  readonly expires_at: string;
  readonly total_size?: number | null;
}

export interface BulkItemResult {
  readonly work_item_id: string;
  readonly outcome: "ok" | "refused" | "skipped";
  readonly code?: string | null;
  readonly detail?: string | null;
}

export interface BulkActionResult {
  readonly action: string;
  readonly succeeded: number;
  readonly refused: number;
  readonly results: readonly BulkItemResult[];
  readonly export_body?: string | null;
  readonly export_mime_type?: string | null;
}

export interface DocumentPreset {
  readonly id: string;
  readonly organization_id?: string | null;
  readonly industry: string;
  readonly document_type: string;
  readonly version: number;
  readonly label: string;
  readonly description: string;
  readonly schema_fields: Record<string, unknown>;
  readonly assertions: readonly { sentence: string; severity: string }[];
  readonly classifier_hints: readonly string[];
  readonly redaction_profile?: string | null;
  readonly is_platform: boolean;
  readonly applied: boolean;
  readonly enabled: boolean;
  readonly safe_harbor_notice?: string | null;
}

export interface RetentionHold {
  readonly id: string;
  readonly workspace_id?: string | null;
  readonly work_item_id?: string | null;
  readonly reason: string;
  readonly reference?: string | null;
  readonly placed_at: string;
  readonly released_at?: string | null;
}

export const createBatch = async (
  workspaceId: string,
  files: readonly BatchFileInput[],
  source: "FILES" | "FOLDER" | "ARCHIVE" = "FILES",
): Promise<Batch> => {
  const response = await apiClient.post<Batch>(
    INGESTION_ENDPOINTS.batches(workspaceId),
    { source, files },
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const getBatch = async (
  workspaceId: string,
  batchId: string,
): Promise<Batch> => {
  const response = await apiClient.get<Batch>(
    INGESTION_ENDPOINTS.batch(workspaceId, batchId),
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const listBatches = async (workspaceId: string): Promise<Batch[]> => {
  const response = await apiClient.get<Batch[]>(
    INGESTION_ENDPOINTS.batches(workspaceId),
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const createUploadSession = async (
  workspaceId: string,
  input: {
    filename: string;
    mime_type?: string;
    total_size?: number;
    sha256?: string;
    batch_item_id?: string;
  },
): Promise<UploadSession> => {
  const response = await apiClient.post<UploadSession>(
    INGESTION_ENDPOINTS.sessions(workspaceId),
    input,
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const getUploadSession = async (
  workspaceId: string,
  sessionId: string,
): Promise<UploadSession> => {
  const response = await apiClient.get<UploadSession>(
    INGESTION_ENDPOINTS.session(workspaceId, sessionId),
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const abortUploadSession = async (
  workspaceId: string,
  sessionId: string,
): Promise<void> => {
  await apiClient.delete(INGESTION_ENDPOINTS.session(workspaceId, sessionId));
};

export const putUploadPart = async (
  workspaceId: string,
  sessionId: string,
  partNumber: number,
  blob: Blob,
  signal?: AbortSignal,
  onUploadProgress?: (event: AxiosProgressEvent) => void,
): Promise<UploadSession> => {
  const response = await apiClient.put<UploadSession>(
    INGESTION_ENDPOINTS.sessionPart(workspaceId, sessionId, partNumber),
    blob,
    {
      headers: { "Content-Type": "application/octet-stream", ...JSON_HEADERS },
      ...(signal ? { signal } : {}),
      ...(onUploadProgress ? { onUploadProgress } : {}),
    },
  );
  return response.data;
};

export const completeUploadSession = async (
  workspaceId: string,
  sessionId: string,
): Promise<{ session_id: string; work_item_id: string; batch_id?: string | null }> => {
  const response = await apiClient.post<{
    session_id: string;
    work_item_id: string;
    batch_id?: string | null;
  }>(INGESTION_ENDPOINTS.sessionComplete(workspaceId, sessionId), undefined, {
    headers: JSON_HEADERS,
  });
  return response.data;
};

/** sha256 of the whole file, so `complete` can refuse a corrupted assembly. */
export const hashFile = async (file: File): Promise<string | undefined> => {
  const subtle = globalThis.crypto?.subtle;
  if (!subtle) {
    // Available only in secure contexts. Without it the server simply has no
    // checksum to compare against, which is a weaker upload, not a broken one.
    return undefined;
  }
  const buffer = await file.arrayBuffer();
  const digest = await subtle.digest("SHA-256", buffer);
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
};

export interface ResumableProgress {
  readonly uploadedBytes: number;
  readonly totalBytes: number;
  readonly partsDone: number;
  readonly partsTotal: number;
}

/**
 * Upload one file through a resumable session.
 *
 * `sessionId` is passed when resuming an upload that a reload interrupted; the
 * server's `parts_received` decides what still has to be sent, so a resumed
 * upload re-sends nothing it does not have to.
 */
export const uploadFileResumable = async (
  workspaceId: string,
  file: File,
  options: {
    sessionId?: string;
    batchItemId?: string;
    signal?: AbortSignal;
    onProgress?: (progress: ResumableProgress) => void;
    onSession?: (sessionId: string) => void;
  } = {},
): Promise<{ sessionId: string; workItemId: string }> => {
  let session: UploadSession;
  if (options.sessionId) {
    session = await getUploadSession(workspaceId, options.sessionId);
  } else {
    const sha256 = await hashFile(file);
    session = await createUploadSession(workspaceId, {
      filename: file.name,
      ...(file.type ? { mime_type: file.type } : {}),
      total_size: file.size,
      ...(sha256 ? { sha256 } : {}),
      ...(options.batchItemId ? { batch_item_id: options.batchItemId } : {}),
    });
  }
  options.onSession?.(session.id);

  const partSize = session.part_size;
  const partsTotal = Math.max(1, Math.ceil(file.size / partSize));
  const held = new Set<number>(session.parts_received);

  for (let partNumber = 1; partNumber <= partsTotal; partNumber += 1) {
    if (options.signal?.aborted) {
      throw new DOMException("Upload cancelled", "AbortError");
    }
    if (held.has(partNumber)) {
      continue;
    }
    const start = (partNumber - 1) * partSize;
    const chunk = file.slice(start, Math.min(start + partSize, file.size));
    await putUploadPart(
      workspaceId,
      session.id,
      partNumber,
      chunk,
      options.signal,
    );
    held.add(partNumber);
    options.onProgress?.({
      uploadedBytes: Math.min(held.size * partSize, file.size),
      totalBytes: file.size,
      partsDone: held.size,
      partsTotal,
    });
  }

  const done = await completeUploadSession(workspaceId, session.id);
  return { sessionId: session.id, workItemId: done.work_item_id };
};

export const runBulkAction = async (
  workspaceId: string,
  input: {
    action: "delete" | "reprocess" | "export" | "tag";
    ids: readonly string[];
    idempotency_key: string;
    tags?: readonly string[];
    export_format?: "csv" | "json";
  },
): Promise<BulkActionResult> => {
  const response = await apiClient.post<BulkActionResult>(
    WORK_ITEM_ENDPOINTS.bulk(workspaceId),
    input,
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const listWorkspaceTags = async (
  workspaceId: string,
): Promise<string[]> => {
  const response = await apiClient.get<{ tags: string[] }>(
    WORK_ITEM_ENDPOINTS.tags(workspaceId),
    { headers: JSON_HEADERS },
  );
  return response.data.tags;
};

export const listPresets = async (
  workspaceId: string,
): Promise<DocumentPreset[]> => {
  const response = await apiClient.get<DocumentPreset[]>(
    INGESTION_ENDPOINTS.presets(workspaceId),
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const applyPreset = async (
  workspaceId: string,
  presetId: string,
): Promise<DocumentPreset> => {
  const response = await apiClient.post<DocumentPreset>(
    INGESTION_ENDPOINTS.presetApply(workspaceId),
    { preset_id: presetId },
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const setPresetEnabled = async (
  workspaceId: string,
  presetId: string,
  enabled: boolean,
): Promise<DocumentPreset> => {
  const response = await apiClient.post<DocumentPreset>(
    INGESTION_ENDPOINTS.presetEnabled(workspaceId, presetId),
    { enabled },
    { headers: JSON_HEADERS },
  );
  return response.data;
};

export const listRetentionHolds = async (
  workspaceId: string,
): Promise<RetentionHold[]> => {
  const response = await apiClient.get<RetentionHold[]>(
    WORK_ITEM_ENDPOINTS.retentionHolds(workspaceId),
    { headers: JSON_HEADERS },
  );
  return response.data;
};

File: backend/app/schemas/ingestion.py
"""ARCH-38 — request and response shapes for batch ingestion."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.storage import DEFAULT_PART_SIZE
from app.services.ingestion.bulk_service import BULK_ACTIONS, MAX_BULK_IDS


class BatchFileRequest(BaseModel):
    """One file the browser intends to upload."""

    client_key: str = Field(min_length=1, max_length=200)
    filename: str = Field(min_length=1, max_length=255)
    size_bytes: int = Field(ge=0)
    sha256: Optional[str] = Field(default=None, pattern=r"^[0-9a-fA-F]{64}$")


class BatchCreateRequest(BaseModel):
    source: Literal["FILES", "FOLDER", "ARCHIVE"] = "FILES"
    files: list[BatchFileRequest] = Field(default_factory=list, max_length=2000)


class BatchItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_key: str
    filename: str
    size_bytes: int
    status: str
    error_code: Optional[str] = None
    error_detail: Optional[str] = None
    work_item_id: Optional[uuid.UUID] = None
    upload_session_id: Optional[uuid.UUID] = None


class BatchResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: str
    source: str
    total_items: int
    completed_items: int
    failed_items: int
    created_at: datetime
    completed_at: Optional[datetime] = None
    items: list[BatchItemResponse] = Field(default_factory=list)


class UploadSessionCreateRequest(BaseModel):
    filename: str = Field(min_length=1, max_length=255)
    mime_type: Optional[str] = Field(default=None, max_length=128)
    total_size: Optional[int] = Field(default=None, ge=0)
    sha256: Optional[str] = Field(default=None, pattern=r"^[0-9a-fA-F]{64}$")
    batch_item_id: Optional[uuid.UUID] = None

    @field_validator("filename")
    @classmethod
    def _strip(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("filename must not be blank")
        return cleaned


class UploadSessionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    filename: str
    part_size: int = DEFAULT_PART_SIZE
    parts_received: list[int] = Field(default_factory=list)
    status: str
    expires_at: datetime
    total_size: Optional[int] = None


class UploadSessionCompleteResponse(BaseModel):
    session_id: uuid.UUID
    work_item_id: uuid.UUID
    batch_id: Optional[uuid.UUID] = None


class BulkActionRequest(BaseModel):
    action: Literal["delete", "reprocess", "export", "tag"]
    ids: list[uuid.UUID] = Field(min_length=1, max_length=MAX_BULK_IDS)
    idempotency_key: str = Field(min_length=8, max_length=128)
    tags: list[str] = Field(default_factory=list, max_length=24)
    export_format: Literal["csv", "json"] = "csv"

    @field_validator("action")
    @classmethod
    def _known(cls, value: str) -> str:
        if value not in BULK_ACTIONS:
            raise ValueError(f"{value!r} is not a bulk action")
        return value


class BulkItemResult(BaseModel):
    work_item_id: str
    outcome: str
    code: Optional[str] = None
    detail: Optional[str] = None


class BulkActionResponse(BaseModel):
    action: str
    succeeded: int
    refused: int
    results: list[BulkItemResult] = Field(default_factory=list)
    export_body: Optional[str] = None
    export_mime_type: Optional[str] = None


class TagListResponse(BaseModel):
    tags: list[str] = Field(default_factory=list)


class PresetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    id: uuid.UUID
    organization_id: Optional[uuid.UUID] = None
    industry: str
    document_type: str
    version: int
    label: str
    description: str
    schema_fields: dict[str, Any] = Field(default_factory=dict)
    assertions: list[Any] = Field(default_factory=list)
    classifier_hints: list[Any] = Field(default_factory=list)
    redaction_profile: Optional[str] = None
    is_platform: bool = True
    applied: bool = False
    enabled: bool = False
    safe_harbor_notice: Optional[str] = None


class PresetApplyRequest(BaseModel):
    preset_id: uuid.UUID


class PresetEnableRequest(BaseModel):
    enabled: bool


class RetentionHoldRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=255)
    reference: Optional[str] = Field(default=None, max_length=128)
    workspace_id: Optional[uuid.UUID] = None
    work_item_id: Optional[uuid.UUID] = None


class RetentionHoldResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: Optional[uuid.UUID] = None
    work_item_id: Optional[uuid.UUID] = None
    reason: str
    reference: Optional[str] = None
    placed_at: datetime
    released_at: Optional[datetime] = None


__all__ = [
    "BatchCreateRequest",
    "BatchFileRequest",
    "BatchItemResponse",
    "BatchResponse",
    "BulkActionRequest",
    "BulkActionResponse",
    "BulkItemResult",
    "PresetApplyRequest",
    "PresetEnableRequest",
    "PresetResponse",
    "RetentionHoldRequest",
    "RetentionHoldResponse",
    "TagListResponse",
    "UploadSessionCompleteResponse",
    "UploadSessionCreateRequest",
    "UploadSessionResponse",
]
